### ISSUE TITLE
Field is not found after migrating from acf-local josn files

### ISSUE THREAD

**Issue by tdwesten** [StoutLogic/acf-builder#61]
Hi,

I'am migrating from local acf-json fields to acf-builder and now ACF can not recognized fields as a repeater field any more. Is gives just a sting with the count of the number of repeater items, like "2". 

(sorry for the unstyled/formatted code but this way is better readeble)

**JSON field config** 
{
            "key": "field_57a989b435b9a",
            "label": "Componisten en Composities",
            "name": "event_composers_composition",
            "type": "repeater",
            "value": null,
            "instructions": "Selecteer hier de componisten en composities die gespeeld worden op\/tijdens dit event. Wanneer een componist of compositie nog niet bestaat kan deze via het plusje toegevoegd worden.",
            "required": 1,
            "conditional_logic": 0,
            "wrapper": {
                "width": "",
                "class": "",
                "id": ""
            },
            "collapsed": "field_57a989da35b9b",
            "min": 1,
            "max": 0,
            "layout": "table",
            "button_label": "Nieuwe regel",
            "sub_fields": [
                {
                    "key": "field_57a989da35b9b",
                    "label": "Componist",
                    "name": "componist",
                    "type": "taxonomy",
                    "value": null,
                    "instructions": "",
                    "required": 1,
                    "conditional_logic": 0,
                    "wrapper": {
                        "width": "",
                        "class": "",
                        "id": ""
                    },
                    "taxonomy": "composer",
                    "field_type": "select",
                    "allow_null": 0,
                    "add_term": 1,
                    "save_terms": 1,
                    "load_terms": 0,
                    "return_format": "id",
                    "multiple": 0
                },
                {
                    "key": "field_57a98a0435b9c",
                    "label": "Compositie",
                    "name": "composition",
                    "type": "taxonomy",
                    "value": null,
                    "instructions": "",
                    "required": 0,
                    "conditional_logic": 0,
                    "wrapper": {
                        "width": "",
                        "class": "",
                        "id": ""
                    },
                    "taxonomy": "composition",
                    "field_type": "select",
                    "allow_null": 1,
                    "add_term": 1,
                    "save_terms": 1,
                    "load_terms": 0,
                    "return_format": "id",
                    "multiple": 0
                }
            ]
        },

**ACF builder config**
 ->addRepeater( 'event_composers_composition', [
                'key' => "field_57a989b435b9a",
                'label' => "Componisten en Composities",
                'instructions' => "Selecteer hier de componisten en composities die gespeeld worden op/tijdens dit event. Wanneer een componist of compositie nog niet bestaat kan deze via het plusje toegevoegd worden.",
                'collapsed' => "componist",
                'layout' => "table",
                'button_label' => "Nieuwe regel",
            ] )
            ->addField( 'componist', 'taxonomy', [
                'key' => "field_57a989da35b9b",
                'label' => "Componist",
                'taxonomy' => "composer",
                'field_type' => "select",
                'add_term' => "1",
                'save_terms' => "1",
                'return_format' => "id",
            ] )
            ->addField( 'composition', 'taxonomy', [
                'key' => "field_57a98a0435b9c",
                'label' => "Compositie",
                'taxonomy' => "composition",
                'field_type' => "select",
                'allow_null' => "1",
                'add_term' => "1",
                'save_terms' => "1",
                'return_format' => "id",
            ] )
            ->endRepeater()

It looks like the key value is not use when registering the fields.

**Comment by stevep** [comment]
@tdwesten You're right, setting keys manually doesn't work. I hadn't considered bringing over existent ACF JSON data, only creating new fields that needed to generate unique keys for. ACF itself uses random strings, while ACF Builder automatically creates keys based upon the field name, and namespaces the keys of sub fields (repeaters and flexible content fields) with their parent field's keys. See: https://github.com/StoutLogic/acf-builder/wiki/generated-keys

The ability to specify a key does sound valuable, but I'm not sure it is a quick fix. For the moment I would recommend continuing to use the JSON definition on existing fields with existing data.

**Comment by tdwesten** [comment]
Damm, we have a lot of old sites using local-json. @jornhenkes and I idd build a importer to migrate these json files to ACF groups. Can wel help in same way and any idea about the timeline?

**Comment by stevep** [comment]
@tdwesten Ah yes @jornhenkes alluded to this project in another issue. So is the code here: https://github.com/StoutLogic/acf-builder/blob/master/src/Transform/NamespaceFieldKey.php#L35-L45

It will prepend the keys of the repeater itself to the subfields. We probably need to check and see if a key has been manually defined on a field first before name spacing it.

Can you do me a favor and paste the output the built code you pasted above? The output of `print_r($builder->build();` I will set up some test cases that we can try to get passing.

**Comment by stevep** [comment]
@tdwesten I'll take a look. Did you make any code changes to get it to work for the key? The NamespaceFieldKey operates on key field and collapsed configuration keys.

**Comment by jornhenkes** [comment]
@stevep Haven't tried to change the code yet, right @tdwesten ? We have some time to work on this together on Monday, so we will keep you informed!

**Comment by tdwesten** [comment]
So I have been thinking about this and looking into the code, I think one of the ways to fix this is when the "key" is set [here](https://github.com/StoutLogic/acf-builder/blob/master/src/FieldBuilder.php#L68) in the FieldBuilder it shut check if the key is not already defined in the $config if this is the case we need add a extra item in the config to define that the key was set in the addField() function. Later on we can use this informatie to prefix only the keys that are not already set.

**Comment by jornhenkes** [comment]
@stevep This is what when ran into as well. We've been testing this last Monday and found the transform function where we got lost a bit. The idea of adding the extra flag to the config sounds like a good solution. That way there's context on the right place. I think using the name _generated_key might be a bit confusing since the code is generating a key as well? But having it in an attribute and would work perfect!

**Comment by stevep** [comment]
@tdwesten @jornhenkes This maybe premature because I want to add more tests, but I wanted to get your feedback first. I pushed an update to the develop branch that adds a [`setCustomKey`](https://github.com/StoutLogic/acf-builder/commit/289e200d47191546bc84899a4d05a1cca6f9da37#diff-0061687a945612312c072ecb61c97adcR140) method to `FieldBuilder` so it should be available on all field types. This will hopefully ensure backwards compatibility, because by not setting it everything works as it was before (though some of the transforms have additional if statements.)

If your using composer in your project set the acf-builder version to dev-develop and give it a whirl. (also make sure min-stability is set to dev) 

Again I still need to add some more tests, but maybe you can test it against your library and see if anything obvious is missing. Check out the [test class](https://github.com/StoutLogic/acf-builder/blob/289e200d47191546bc84899a4d05a1cca6f9da37/tests/FieldsBuilderCustomFieldKeysTest.php) to see how its being used.

**Comment by stevep** [comment]
Damn this is premature because it doesn't work in PHP < 7. It dies in [RecursiveTransform::transform](https://github.com/StoutLogic/acf-builder/blob/289e200d47191546bc84899a4d05a1cca6f9da37/src/Transform/RecursiveTransform.php) when $config is modified inside the array walk. 

```php
array_walk($config, function (&$value, $key) use (&$config) {
  ...
}
```

**Comment by stevep** [comment]
Phew solved the array_walk issue by replacing it with a simple foreach, build passes in php < 7 now.  Anyway, try it now and let me know how it goes.

**Comment by jornhenkes** [comment]
Great, we're gonna give it a try today!

**Comment by LukeAbell** [comment]
@tdwesten Any chance you could publish that importer you created? Would love to use it on our installs :)
